pkg/collector/runner/runner.go

```diff
@@ -69,6 +90,46 @@ func NewRunner(numWorkers int) *Runner {
 	return r
 }
 
+// UpdateNumWorkers checks if the current number of workers is reasonable, and adds more if needed
+func (r *Runner) UpdateNumWorkers(numChecks int64) {
+	numWorkers, _ := strconv.Atoi(runnerStats.Get("Workers").String())
+
+	if r.staticNumWorkers {
+		return
+	}
+
+	// Find which range the number of checks we're running falls in
+	var desiredNumWorkers int
+	switch {
+	case numChecks <= 10:
+		desiredNumWorkers = 4
+	case numChecks <= 15:
+		desiredNumWorkers = 10
+	case numChecks <= 20:
+		desiredNumWorkers = 15
+	case numChecks <= 25:
+		desiredNumWorkers = 20
+	default:
+		desiredNumWorkers = maxNumWorkers // max = 25
+	}
+
+	// Add workers if we don't have enough for this range
+	added := 0
+	for {
+		if numWorkers >= desiredNumWorkers {
```

Comment:
Wondering if scaling down has an actual real-world use-case. Normally checks will be configured once and remain pretty static throughout the duration of the agent's life. That said, because of _AutoDiscovery_ there are cases where the number of configured checks could maybe go down(?). The overhead in memory for the additional go-routines is quite small, so I'm not sure how badly we'd want to implement the down-sizing.... It could maybe be implemented with a channel, or an atomic variable in the `Runner` class for the number of workers that should shutdown. We could discuss if you feel this makes sense.

Reply:
That would be a nice feature, but I think for it to be really useful we would need there to be cases where there's at least 10 checks that get cancelled (otherwise the difference would be <=5 workers... not very significant), and I don't think that would happen often in an agents lifetime... not sure if it's worth implementing now? Definitely an option to keep in mind though!